3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/UaParserEnrichment.scala

```diff
@@ -33,34 +37,87 @@ import ua_parser.Client
 
 // json4s
 import org.json4s._
+import org.json4s.DefaultFormats
 import org.json4s.JValue
 import org.json4s.JsonDSL._
 import org.json4s.jackson.JsonMethods._
 
 // Iglu
 import iglu.client.{SchemaCriterion, SchemaKey}
+import iglu.client.validation.ProcessingMessageMethods._
 import utils.ScalazJson4sUtils
 
 /**
  * Companion object. Lets us create a UaParserEnrichment
  * from a JValue.
  */
 object UaParserEnrichmentConfig extends ParseableEnrichment {
+  implicit val formats = DefaultFormats
 
   val supportedSchema = SchemaCriterion("com.snowplowanalytics.snowplow", "ua_parser_config", "jsonschema", 1, 0)
 
-  def parse(config: JValue, schemaKey: SchemaKey): ValidatedNelMessage[UaParserEnrichment.type] =
-    isParseable(config, schemaKey).map(_ => UaParserEnrichment)
+  private val localRulefile = "./ua-parser-rules.yml"
+
+  def parse(config: JValue, schemaKey: SchemaKey): ValidatedNelMessage[UaParserEnrichment] = {
+    val c = UaParserEnrichment(None)
+    isParseable(config, schemaKey).flatMap(conf => {
+      (for {
+        rules <- getCustomRules(conf)
+      } yield UaParserEnrichment(rules)).toValidationNel
+    })
+  }
+
+  private def getCustomRules(conf: JValue): ValidatedMessage[Option[(URI, String)]] =
+    if (ScalazJson4sUtils.fieldExists(conf, "parameters", "regexFileUri")) {
+      for {
+        rulefile <- ScalazJson4sUtils.extract[String](conf, "parameters", "regexFileUri")
+        source   <- getUri(rulefile)
+      } yield (source, localRulefile).some
+    } else {
+      None.success
+    }
+
+  private def getUri(uri: String): ValidatedMessage[URI] =
+    ConversionUtils
+      .stringToUri(uri)
+      .flatMap(_ match {
+        case Some(u) => u.success
+        case None    => "A valid URI to ua-parser regex file must be provided".fail
+      })
+      .toProcessingMessage
 }
 
 /**
  * Config for an ua_parser_config enrichment
  *
  * Uses uap-java library to parse client attributes
  */
-case object UaParserEnrichment extends Enrichment {
+case class UaParserEnrichment(customRulefile: Option[(URI, String)]) extends Enrichment {
+
+  override val filesToCache: List[(URI, String)] =
+    customRulefile.map(List(_)).getOrElse(List.empty)
+
+  lazy val uaParser = {
+    def constructParser(input: Option[InputStream]) = input match {
+      case Some(is) =>
+        try {
+          new Parser(is)
+        } finally {
+          is.close()
+        }
+      case None => new Parser()
+    }
 
-  val uaParser = new Parser()
+    val parser = for {
+      input <- Validation.fromTryCatch(customRulefile.map(f => new FileInputStream(f._2)))
+      p <- try {
+        constructParser(input).success
```

Comment:
Cannot it also be `fromTryCatch`?

Reply:
actually I think we shouldn't use `fromTryCatch` at all since it also catches fatal errors in the version of scalaz we use

Reply:
Good point.